[tcp] prepare for TCP implementation

samkumar opened the pull request:
This is the third in a series of pull requests (after #6744 and #6770) bringing TCPlp to OpenThread. I'm making a series of small pull requests so that it's easier for the OpenThread community to review the code. If you'd like to preview a working version of TCPlp in OpenThread, see #6650.

This pull request consists of three commits that add functionality in OpenThread needed for TCP and performs a small amount of initialization/deinitialization work in TCP itself. The purpose of this is to prepare for implementing TCP logic in future pull requests.

samkumar commented:
I noticed that some of the CI tests are failing due to warnings related to how I handled timers. I would appreciate some guidance from the OpenThread community on how best to place timers in the `otTcpEndpoint` structure.

samkumar commented:
I just rebased this onto `main` to facilitate code review.

abtink reviewed (commented):
Thank you @samkumar.

Added some questions/suggestions below.

samkumar commented on src/core/net/tcp6.hpp:
```diff
@@ -311,6 +310,36 @@ class Tcp : public InstanceLocator, private NonCopyable
+            uint8_t *ptr = reinterpret_cast<uint8_t *>(&aTimer);
+            ptr -= offsetof(Endpoint, mTimers[aIndex].mSize[0]);
+            return *reinterpret_cast<Endpoint *>(ptr);
+        }
```
I agree with the drawbacks you've identified. We had a discussion earlier about the question of having OpenThread maintain a pool, and we had decided not to do that. The deciding argument being that maintaining a pool in OpenThread would require OpenThread to be re-compiled if the pool size is changed, but that's undesirable in certain situations.

I like the idea you suggested below, of having a single timer in the TCP class. My understanding is that, on expiry, it would iterate over all TCP endpoints and check the four timers on each one to check which ones expired. And when a timer expires or is cancelled, it would iterate over all TCP endpoints to find the timer that expires next. That would be inefficient when there is a large number of TCP endpoints, but it will probably OK since we only expect there to be few TCP endpoints.

Now, we could improve on that (decreasing computation time but increasing memory cost) by having intrusive doubly-linked lists in each TCP endpoint. That way, we can maintain a sorted list for each timer, and avoid having to iterate over the whole linked list. But this is getting into the territory of what `Timer` is support to handle, and it's a bit disappointing that `Timer` is proving unsuitable for this. I'm wondering whether the best thing to do would be to change `Timer` to make it suitable --- for example, remove the inheritance from `InstanceLocator` and provide a C struct counterpart to include in `otTcpEndpoint`. What are your thoughts on doing this?

abtink replied:
> I agree with the drawbacks you've identified. We had a discussion earlier about the question of having OpenThread maintain a pool, and we had decided not to do that. The deciding argument being that maintaining a pool in OpenThread would require OpenThread to be re-compiled if the pool size is changed, but that's undesirable in certain situations.

I know we discussed this and I do see that you already spent the time to find a solution/code to make it work but I'd still think it will be good for us to re-visit it.

I think different solutions may address all situations, and longer term having `TcpEndPoint` be an internal definition (i.e., not having it as `otTcpEndPoint` in public header) will help simplify things.

The key drawback for me (in addition to complexity it adds) is that in this model, we may end up over-estimating the size (use larger than necessary buffers for `otTcpEndPoint`) which is not a good for memory-contained devices.

I don't want us to penalize the memory-constrained case in favor of flexibility for the not-constrained case.

Here is my view on this:

- On a memory-constrained device, one way or another, the app developer needs to determine the the max number of TCP end points that may be needed. I think this would apply independent of whether it is being managed by app layer code or we use a pool and have an OT configuration to specify the pool size.
   - And as I said there are already other items (like max number of unicast/multicast address) that have a hard max limit which is determined by an  `OPENTRHEAD_CONFIG` and if the app need changes, we need to re-build OT stack.

- If we need flexibility and support for dynamic number of end-points (which I guess makes sense on a not-very-memory constrained platform), then
  1) We can still use a pool with large enough estimate for number of entries that cover the application use-cases.
      - Even if the pool size needs a change, requiring a re-build of the OT stack should not add a lot of overhead.
  2) We can have an option (under certain `OPENTHREAD_CONFIG`) where OT stack uses heap to allocate the `TcpEndPoint` instance instead of having a pool.
      - This way it is more dynamic and tied to platform (heap size available)
      - For example, this makes sense for POSIX like platform where we can have way more memory available and can use heap.
   3) We can also consider adding a third option (again under some `OPENTHREAD_CONFIG`) where we still require the user to provide the buffer for the TCP end points but instead of defining the `otTcpEndPoint` struct in public API, we have an API where the user can ask and get the size of the buffer needed for a TCP end-point object. Then we expect the caller to provide a buffer of at least that size to us (which we then use `new` to convert to a `TcpEndPoint` instance).

Thoughts?

samkumar replied:
I don't mind revisiting this issue. I don't think it would be much work to change the API to use a pool-based approach, if that's what we decide to do. It would be great if @jwhui and @simonlingoogle could share their thoughts as well, as they participated in previous discussions of this point.

There are a few things I'd like to point out though:
* I understand the argument of possibly wasting memory (4 B per timer). But more memory is wasted simply by using a `Timer` object for each timer in a TCP endpoint. This is because (1) each timer has its own `mNext` pointer, (2) each timer has its own `mHandler` pointer, and (3) each timer potentially has a pointer to its enclosing otInstance. That's 8-12 bytes of overhead per timer. Adding 0-4 bytes due to overestimating the size of a TCP timer seems small in comparison. If we want to really save RAM, we should take the approach of having a single global timer, and multiplexing all the TCP timers on top of it. That way, we can share the same linked list for all timers (sorted in order of each TCP endpoint's earliest timer). In my opinion, this is the best solution.
* In my view, the root cause of the problem is not exposing `otTcpEndpoint` in C. Rather, it is the fact that we can't put a `Timer` in a C struct. I think it's reasonable to expect this to work because other RTOSes I've used (TinyOS and RIOT OS) allow it. So, finding a way to expose timers to C applications is probably the best way to address the root cause of this (though it would still be less efficient than what's proposed above). All we would need to do to provide a version of `TimerMilli` that doesn't inherit from `InstanceLocator`.

I think there are advantages to having applications allocate `otTcpEndpoint`s directly. I also think that there are advantages to having a pool of them allocated by OpenThread. However, I don't think that the issue we're seeing here with `Timer`s is a good reason to switch, for the two reasons I mentioned above.

Thoughts? @abtink @simonlingoogle @jwhui

abtink replied:
> the root cause of the problem is not exposing otTcpEndpoint in C. Rather, it is the fact that we can't put a Timer in a C struct.

This is now happening with `Timer` but can happen with other internal types and definitions. By having a C public `otTcpEndpoint` definition, we basically limit what we can include in it. Whereas if it is opaque public type and we define it  `class TcpEndPoint` as internal C++ type, we can use all other types/definitions from OT in it. For example,  we can have it inherit from `InstanceLocator` (which then optimizes how `mInstance` is stored), or use `Tasklet`, `LinkedList<>` or even other helper types (say `RssAverager`).

In terms of design, I think if the user does not need to know/access certain info, it's better/safer to hide the info. We don't intend the user of TCP API to access `otTcpEndPoint` fields anyway, so not having a definition of this struct in  in public C header makes it a simpler model. The only thing we really care is for user to know the size (if we expect user to provide buffer for its instances which we can do from an API).

> That's 8-12 bytes of overhead per timer. Adding 0-4 bytes due to overestimating the size of a TCP timer seems small in comparison. 

Yes! That's the reason I suggested remembering the expire time and sharing a `Timer` (insead of multple timers)

>  and multiplexing all the TCP timers on top of it. That way, we can share the same linked list for all timers (sorted in order of each TCP endpoint's earliest timer)

This is actually somewhat similar to what `Timer` implementation does!... :) 

> I think it's reasonable to expect this to work because other RTOSes I've used (TinyOS and RIOT OS) allow it. So, finding a way to expose timers to C applications is probably the best way to address the root cause of this (though it would still be less efficient than what's proposed above)

The fact that we don't expose `Timer` or `Tasklet` (or other OS/platform related functions) in OT public APIs is very much intentional. We do not want OpenThread to act as or replace an OS.  OT is intended as Thread stack implementation which is portable and can run on different OSes and platforms.

samkumar replied:
Thanks @abtink for clarifying this. I hadn't previously realized that it was recommended to avoid using a large number of `Timer`s, though in retrospect, this makes total sense.

I had realized that sharing a linked list for all timers is how `Timer` and `TimerScheduler` worked; the reason I had originally tried to instantiate `Timer`s within the TCP endpoints was to avoid reimplementing this. But based on what you've said above, I understand why creating that many `Timer`s is a bad idea.

Here's a concrete proposal for moving forward with this. I'll rewrite the timer code to have a single `Timer` in the TCP class. This seems to be the best option since we both agree that this is the most memory-efficient.

Meanwhile, we'll leave the question as to whether to have a pool of `otTcpEndpoint`s open for additional discussion; hopefully @jwhui and @simonlingoogle will chime in. I don't want merging in the TCP implementation to stall while we discuss what to do about the `otTcpEndpoint` API, since it sounds like both can proceed independently for now.

Does that sound like a good plan?

abtink replied:
Sounds good to me. Thanks. 👍

jwhui replied:
> Meanwhile, we'll leave the question as to whether to have a pool of `otTcpEndpoint`s open for additional discussion; hopefully @jwhui and @simonlingoogle will chime in. I don't want merging in the TCP implementation to stall while we discuss what to do about the `otTcpEndpoint` API, since it sounds like both can proceed independently for now.

OpenThread core hardcoding the size of the pool requires recompiling the library whenever the pool size needs to change. One concern with recompiling is that it does not allow Thread Group's certification-by-inheritance, as the component-level certification would no longer applied to a recompiled library.

abtink replied:
How about this model from https://github.com/openthread/openthread/pull/6790#discussion_r683057291?

>  we still require the user to provide the buffer for the TCP end points but instead of defining the `otTcpEndPoint` struct in public API, we have an API where the user can ask and get the size of the buffer needed for a TCP end-point object. Then we expect the caller to provide a buffer of at least that size to us (which we then use new to convert to a TcpEndPoint instance).

It lets us hide the definition in `TcpEndPoint` while being flexble and requireing the user to manage the buffer/space for them. The only downside is it puts the burdon on caller to get the size from API call and provide buffer of at least that size (if OT is built as a library and linked with different app layer code, the returned value from API (i.e. the size) would be fixed/constant).

simonlingoogle replied:
A pool TCP sockets of works for me. 

#6036 defined the opaque TCP structure to address this problem, but it's tricky to get the size right.

```
struct otTcpSocket
{
#if __SIZEOF_POINTER__ == 8
    uint64_t mMemHolder[35];
#elif __SIZEOF_POINTER__ == 4
    uint32_t mMemHolder[50];
#else
#error "only 32bit and 64bit are supported"
#endif
};
```

samkumar commented on src/core/net/tcp6.hpp:
```diff
@@ -311,6 +310,36 @@ class Tcp : public InstanceLocator, private NonCopyable
+        TimerMilli &     GetKeepTimer() { return GetTimer(2); }
+        static Endpoint &FromKeepTimer(TimerMilli &aTimer) { return FromTimer(aTimer, 2); }
+
+        TimerMilli &     Get2MslTimer() { return GetTimer(3); }
```
I like this approach, and I think it is the one we should go for, for now. See also my more detailed response to your previous comment.

abtink replied:
> My understanding is that, on expiry, it would iterate over all TCP endpoints and check the four timers on each one to check which ones expired. And when a timer expires or is cancelled, it would iterate over all TCP endpoints to find the timer that expires next. 

My suggestion is share one timer per end-point (instead of 4 timers per end point).

samkumar replied:
I see. If we do that, we'd still have to find a way to put one `Timer` object in each TCP endpoint though.

What are your thoughts on using a single `Timer` for all TCP endpoints? The advantage there is that the timer would not be in the `otTcpEndpoint` struct. This would eliminate the problems you described earlier and the drawbacks of the solution that I found.

abtink replied:
I was not considering one `Timer` for all TCP end point but I agree that can also work. :)

As you said the challenge is its scalability. I would guess if we have smaller number of TCP end points, (say in range 10s-20s end points) then maybe iterating over all would be fine.

samkumar replied:
Regarding scalability, the current `Timer` implementation also has to deal with this to some extent. Although it uses a sorted linked list, I believe it does require iterating over the linked list to add new timers or cancel timers. We can also keep the existing Endpoint linked list sorted in order of each endpoint's first expiring timer.

abtink replied:
Exactly! `Timer::Scheduler` keeps track of sorted linked list of all scheduled `Timer` objects and may need to iterate through the list from some methods. We don't want to have too many timers (this is basically limited by the number of modules in OT).
- As example, we added `TimeTicker` (from https://github.com/openthread/openthread/pull/5379) which helped us remove a bunch of timers and share one among a bunch of modules.

samkumar replied:
OK. Thanks for the context that OpenThread tries to avoid having too many `Timer`s.

simonlingoogle replied:
Agree with @abtink . Actually, #6036 uses one Timer for all TCP sockets.

samkumar replied:
The latest update to the code below also uses one Timer for all TCP sockets. Thanks again for the guidance here.

samkumar commented:
@abtink I re-implemented timers in #6650 as we discussed and I have updated this pull request accordingly. Additionally, I added some code that makes the IPv6 header available to the (stub for the) TCP processing routine for received segments. When you get a chance, please take another look at this.

simonlingoogle commented on include/openthread/tcp.h:
```diff
@@ -215,6 +215,8 @@ struct otTcpEndpoint
     otTcpReceiveAvailable mReceiveAvailableCallback; ///< "Receive available" callback function
     otTcpDisconnected     mDisconnectedCallback;     ///< "Disconnected" callback function
 
+    uint32_t mTimers[4];
```
remove hardcoding 4?

simonlingoogle replied:
```suggestion
    TimeMilli mTimers[4];
```

samkumar replied:
I agree with @abtink. `TimeMilli` is nothing more than a wrapper around `uint32_t` with some methods added. So I just construct the `TimeMilli`s as I need them in the timer implementation.

simonlingoogle commented on src/core/net/ip6.cpp:
```diff
@@ -452,7 +452,14 @@ Error Ip6::SendDatagram(Message &aMessage, MessageInfo &aMessageInfo, uint8_t aI
     uint16_t payloadLength = aMessage.GetLength();
 
-    header.Init();
+    if ((aMessageInfo.mVersionClassFlow >> 28) == 0x6)
```
avoid hardcoding (28, 0x6) ?

simonlingoogle commented on src/core/net/ip6.cpp:
```diff
@@ -963,7 +975,7 @@ Error Ip6::HandlePayload(Message &          aMessage,
 #if OPENTHREAD_CONFIG_TCP_ENABLE
     case kProtoTcp:
-        error = mTcp.ProcessReceivedSegment(*message, aMessageInfo);
+        error = mTcp.ProcessReceivedSegment(aIp6Header, *message, aMessageInfo);
```
```suggestion
        error = mTcp.HandleMessage(aIp6Header, *message, aMessageInfo);
```

Keep the naming consistent with UDP and ICMP6?

simonlingoogle commented on src/core/net/tcp6.cpp:
```diff
@@ -37,28 +37,49 @@
+    mReceiveAvailableCallback = aArgs.mReceiveAvailableCallback;
+    mDisconnectedCallback     = aArgs.mDisconnectedCallback;
+
+    mInstance = &aInstance;
```
Instead of storing `mInstance` to Endpoint, can we make Endpoint a subclass of InstanceLocator?

samkumar replied:
That wouldn't work, since it corresponds to the `otTcpEndpoint` class in the public API.

simonlingoogle commented on src/core/net/tcp6.cpp:
```diff
@@ -37,28 +37,49 @@
+
+    mInstance = &aInstance;
+
+    error = aInstance.Get<Tcp>().mEndpoints.Add(*this);
```
Can this fail?

samkumar replied:
No, it should never fail. It would only fail if this endpoint were already part of the linked list, which should never happen.

samkumar replied:
Actually, I looked into this more, and I realized that `Add` actually checks if the element is already in the linked list. So I'm going to change this to do this check first (to protected against double-initialization) and exist if it's already part of the linked list.

simonlingoogle commented on src/core/net/tcp6.cpp:
```diff
@@ -136,20 +157,121 @@ Error Tcp::Endpoint::Abort(void)
+    Error     error;
+    Endpoint *prev;
+
+    Tcp &tcp = GetInstance().Get<Tcp>();
+    SuccessOrExit(error = tcp.mEndpoints.Find(*this, prev));
```
```suggestion
    Tcp &tcp = GetInstance().Get<Tcp>();

    SuccessOrExit(error = tcp.mEndpoints.Find(*this, prev));
```

simonlingoogle commented on src/core/net/tcp6.cpp:
```diff
@@ -136,20 +157,121 @@ Error Tcp::Endpoint::Abort(void)
+    {
+        if (IsTimerActive(timerIndex))
+        {
+            Time expiry(mTimers[timerIndex]);
+            hasUnfiredTimer            = true;
```
```suggestion
            Time expiry(mTimers[timerIndex]);

            hasUnfiredTimer            = true;
```

simonlingoogle commented on src/core/net/tcp6.cpp:
```diff
@@ -136,20 +157,121 @@ Error Tcp::Endpoint::Abort(void)
+
+bool Tcp::Endpoint::GetEarliestActiveTimer(TimeMilli aNow, uint32_t &aDelayUntilNextActiveTimer)
+{
+    bool hasUnfiredTimer       = false;
+    aDelayUntilNextActiveTimer = Timer::kMaxDelay;
```
```suggestion
    bool hasUnfiredTimer       = false;

    aDelayUntilNextActiveTimer = Timer::kMaxDelay;
```

simonlingoogle reviewed (approved):
SGTM overall. 

Some code is incomplete thus not easy to fully review.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -136,20 +157,121 @@ Error Tcp::Endpoint::Abort(void)
+            hasUnfiredTimer            = true;
+            aDelayUntilNextActiveTimer = OT_MIN(aDelayUntilNextActiveTimer, expiry - aNow);
+        }
+    }
```
There are some helper methods in `Time` class which may help here.

The `Time` class overloads the `operator<()` to handle the comparison even when actual time may roll-over (this is the main purpose of it 😃...)

We also have `now.GetDistantFuture()` and `now.GetDistantPast()` which we can then use to safely compare against a bunch of other `Time` instances to find the earliest or latest time.

As example, you can check `Commissioner::UpdateJoinerExpirationTimer()` in `commissioner.cpp` on how they are used (or `Client::HandleTimer()` in `dns_clinet.cpp`).

abtink replied:
There is also `TimerMilli::FireAtIfEarlier()` which is helpful in many situations (in addition to simpler `TimerMilli::FireAt()`).

```c++
   /**
     * This method (re-)schedules the timer with a given a fire time only if the timer is not running or the new given
     * fire time is earlier than the current fire time.
     *
     * @param[in]  aFireTime  The fire time.
     *
     */
    void FireAtIfEarlier(TimeMilli aFireTime);
```

samkumar replied:
Thanks. I'm aware of `Time` and `TimeMilli`. But in this case, I'm using `OT_MIN` to compare delays (delay = difference between two times), not times. Looking at the time library, it looks like delays are represented as `uint32_t` rather than a dedicated class, and comparing them is done using the default `<` operator rather than an overloaded one.

I suppose I can re-work this code to just compute the `Time` of expiry rather than a delay starting at the current time. Would that be preferable?

samkumar replied:
I went ahead and refactored this to work with absolute times instead of delays and use `FireAt`/`FireAtIfEarlier`.

samkumar commented on src/core/net/tcp6.cpp:
```diff
@@ -37,28 +37,49 @@
 
 #include "tcp6.hpp"
 
+#include <new>
```
This is left over from the previous version, which used placement new. I can probably delete this now.

samkumar commented on include/openthread/ip6.h:
```diff
@@ -213,14 +213,17 @@ typedef struct otSockAddr
+    const void * mLinkInfo; ///< A pointer to link-specific information.
+    uint32_t
+        mVersionClassFlow; ///< Version, class, and flow information. If version is not 6 (0b110), a default is used.
+                           ///< DSCP bits of this field are overridden by the priority attached to the otMessage.
```
Sorry for missing this comment earlier. The purpose of this is to allow one to process IP packets in an ECN-aware way --- to mark packets as ECN-capable when sending them out, and checking if they were marked as "congestion experienced" when receiving them.

The `MessageInfo` class is tied to `otMessageInfo`, so adding the field required changing it in the public API.

This is primarily to be able to read and write the traffic class --- specifically the last two bits of the traffic class, corresponding to ECN. I bundled it with the version and class since that's how the IPv6 implementation dealt with it. Should I just keep it to specifying ECN for now? Or should I split the DCSP, ECN, and Flow bits into separate fields?

samkumar replied:
The latest update changes this to just have a field `mEcn` to avoid redundantly specifying DCSP and version. I figured that if we ever need to specify the flow, we can add `mFlow` at that time. I hope this addresses your concern that bundling these together into a single field might be confusing for the programmer.

samkumar commented on src/core/net/tcp6.hpp:
```diff
@@ -311,6 +310,36 @@ class Tcp : public InstanceLocator, private NonCopyable
+        static Endpoint &FromTimer(TimerMilli &aTimer, size_t aIndex)
+        {
+            uint8_t *ptr = reinterpret_cast<uint8_t *>(&aTimer);
+            ptr -= offsetof(Endpoint, mTimers[aIndex].mSize[0]);
```
(Looks like I missed this too)

OK. With the timer redesign that we agreed on after you posted this, this won't be necessary anymore.

samkumar commented on src/core/common/logging.hpp:
```diff
@@ -81,6 +81,7 @@ extern "C" {
+#define _OT_REGION_TCP_PREFIX "-TCP-----: "
 #define _OT_REGION_MAC_PREFIX "-MAC-----: "
 #define _OT_REGION_MEM_PREFIX "-MEM-----: "
 #define _OT_REGION_NCP_PREFIX "-NCP-----: "
```
Thanks for the pointer --- I'll make this change.

abtink replied:
Thanks.

samkumar commented:
I just pushed a new version of the code. I believe all comments so far are addressed. @simonlingoogle @abtink @jwhui Want to take another look to make sure it's OK now?

samkumar commented:
I also rebased on the latest `main` branch by the way.

abtink reviewed (commented):
Thanks @samkumar . Looks good overall.

Some suggestions below which hopefully help simplfiy the handling of timers.

samkumar commented on src/core/net/tcp6.cpp:
```diff
@@ -136,20 +156,130 @@ Error Tcp::Endpoint::Abort(void)
+                 static_cast<unsigned int>(aDelay));
+
+    GetInstance().Get<Tcp>().ResetTimer(false);
+}
```
If we make the other change you suggested (giving up the invariant that `mTimer`'s expiry matches the earliest timer), then this works.

samkumar commented on src/core/net/tcp6.cpp:
```diff
@@ -136,20 +156,130 @@ Error Tcp::Endpoint::Abort(void)
+    if (GetInstance().Get<Tcp>().mTimer.GetFireTime() == TimeMilli(mTimers[timerIndex]))
+    {
+        GetInstance().Get<Tcp>().ResetTimer(false);
+    }
```
This works too. I updated the code according to your suggestion.

The trade-off is that `mTimer` will expire more often. In contrast, with the current code, it only expires when a TCP timer expires. For context, TCP will set its retransmit timer every RTT, so now we will have `mTimer` expiring once every RTT, even in the common case where the ACK arrives and TCP resets the timer.  The trade-off is what you consider to be expensive.  In previous platforms I've worked with, waking up the CPU from sleep was expensive, so I wrote the timer code to avoid expiring the timer unless we absolutely had to. For OpenThread it's unclear, so I went with your simpler alternative for now. It looks like the default alarm driver wakes up the CPU periodically anyway so it won't make a difference. But if the driver is ever changed to manage the timer in a tickless way, it might be worth revisiting this.

samkumar replied:
To clarify, this is a trade-off between two types of overhead: spurious wakeups from sleep, and additional mutations to the linked list of timers. The old method, which maintained the invariant the `mTimer` is set to the expiry time of the earliest TCP timer requires fewer wakeups from sleep, but requires more linked list mutations to maintain that invariant.

If later on, we decide that we want to minimize the number of times `mTimer` fires for performance reasons, at the expense of mutating the linked list more frequently by setting and cancelling `mTimer`, the best way to do that is to add a single call to `ResetTimer` after processing each input packet. That way, traversing the linked list of endpoints can happen once per input packet, not once each time TCPlp sets a timer (which happens multiple times per input packet). I've added a comment in the code to this effect, in case we revisit this point earlier.

samkumar commented on src/core/net/tcp6.cpp:
```diff
@@ -136,20 +156,130 @@ Error Tcp::Endpoint::Abort(void)
+            TimeMilli expiry(mTimers[timerIndex]);
+
+            hasUnfiredTimer = true;
+            aExpiryTime     = OT_MIN(aExpiryTime, expiry);
```
I think the current code is correct.

Even in the previous version, I think that it would have been safe. If you call `Timer::FireAt` with a time that's in the past (`< now` as compared in the timer code) then the alarm is set to 0 msec and the timer tasklet will be posted immediately. See the code here: https://github.com/openthread/openthread/blob/main/src/core/common/timer.cpp#L173. So there's no harm in calling `GetEarliestActiveTimer()` even if there are expired timers, since you will choose an expired timer, set  `mTimer` to fire at its expired time in the past, causing the alarm to be set to 0 msec, and then call those callbacks when the Timer callback occurs.

I did implement your suggestion for the reasons described in the comment above (simplicity and overhead of linked list manipulations vs extra wakeups). So I believe the current code is correct, regardless of my explanation in the previous paragraph.

abtink replied:
Yes. The `FireAt()` works correctly if the fire time is in the past (that's very much intentional in the timer design and we use and rely on it often 😃).

But that was not the place that could potentially cause an issue if we call `GetEarliestActiveTimer()` with expired timers in the previous code. The issue could happen on `OT_MIN()` when we try to keep track of the next earliest fire time.
- We init the `aExpiryTime = aNow.GetDistantFuture()`.
- By definition `aNow.GetDistantFuture()` gives a time that is ahead of any other time that is ahead of `aNow`. 
- If we now compare `aExpireTime` with a time that is past of `aNow` (expired timer) that won't work correctly.
- Note that the same situation can happen if we don't use `GetDistantFuture()` and instad of try to track the delay duration (i.e. if we try to get delay till next expire time using `(aExpiryTime - aNow)` we need to check and ensure it is ahead of `aNow`).

We need to either ensure expired times are handled  or update code to check for this case as we try to find next earliest fire time.

samkumar replied:
OK, thanks for clarifying. I hadn't thought about that. I think the previous version probably had this bug, but the latest version doesn't have it. That's because there are no `ResetTimer(false)` calls in the latest version, and I believe that `GetEarlisetActiveTimer()` will never be called when there are pending timers.

samkumar replied:
To make this even more clear, I actually combined `GetEarliestActiveTimer` and `FirePendingTimers` in the latest revision. I realized that, with `SetTimer` now calling `FireAtIfEarlier`, we no longer need to scan the list of timers in each TCP endpoint twice.

abtink replied:
Sounds good to me. Thanks. 😃

samkumar commented on src/core/net/tcp6.cpp:
```diff
@@ -166,15 +296,98 @@ Error Tcp::Listener::StopListening(void)
+     * anyway reset mTimer after firing the callbacks. The mFiringTimers
+     * variable is how this is enforced.
+     */
+    if (aFirePendingTimers)
```
I think that the current version of the code is fine. The timers callbacks don't ever call `ResetTimer()` anymore, so don't have to worry about `mTimer` being mutated by the callbacks. They may change other timer values on the same TCP endpoint, but a timer callback in one TCP endpoint won't ever affect timers on another TCP endpoint. So we can just iterate over the linked list once. For each endpoint, we iterate over its timers twice: once to fire any callbacks, and a second time to determine the earliest timer.

abtink replied:
I think better to use `FireAtIfEarlier()` at the end. I see you are using `FiretAt()` at end of `FireAllPendingTimers()` method.

When handling expired timers, we can invoke callbacks and I think potentially even go out of OT code (callback to user and app layer code) which can then call any other APIs. When calling a callback we should assume things may change after we return from it.

Another related point: when iterating over end points `for (Endpoint *endpoint = mEndpoints.GetHead(); endpoint != nullptr; endpoint = endpoint->GetNext())` we are assuming `endpoint` remains valid and on list (so we can use `endpoint->GetNext()` to get to the next on the list. Can it be possibly removed from one of the timer callbacks?

If so, we need to handle this in the code (one way to do this when iterating over linked-list is to save `next = endpoint->GetNext()` at the top of the loop in a variable and use `endpoint = next` in loop iteration step, e.g see `MeshForwarder::UpdateReassemblyList()`). But there may be other situations to consider after calling `endpoint->FirePendingTimers(now);` (if it can change `endpoint`?).

samkumar replied:
The TCPlp timer callbacks are TCP functions. Most user-specified callbacks are called in response to receiving a packet, which is why I hadn't thought about this earlier... but thinking about what you said, there is one callback--- `connection_lost`---that might be called in response to a timer event, if it causes the whole TCP connection to time out.

So I agree that we need to write this more defensively, since the callback might enter user code, which may update timers on the same TCP endpoint and/or the other endpoints, or even mutate the linked list. This is what I think we can assume:
* Any newly set timers won't have expired yet, since TCPlp will only set timers to expire in the future (including some minimum delay)
* Any newly created TCP endpoints won't have any expired timers (corollary of above), but they might have active timers (if the user code ended up calling `otTcpEndpointConnect` or something).

I think that either of the two approaches you suggested (the first approach is to `FireAtIfEarlier`, the second approach is to iterate over the `Endpoint`s twice) would work. For now, I've just used `FireAtIfEarlier` since that's more efficient.

samkumar commented on src/core/net/tcp6.cpp:
```diff
@@ -136,20 +156,130 @@ Error Tcp::Endpoint::Abort(void)
+    Tcp &tcp = GetInstance().Get<Tcp>();
+
+    SuccessOrExit(error = tcp.mEndpoints.Find(*this, prev));
+    tcp.mEndpoints.PopAfter(prev);
```
I think having the error is fine. I updated it as you suggested.

samkumar commented:
@abtink I updated the code again. Want to take another look?

I simplified the timers as you suggested --- never cancelling `mTimer`, and just let it expire and do nothing if there's no longer a TCP timer at its original expiry time. As I explained in the comments above, it's really a performance tradeoff (extraneous timer expiry events vs. extra scans of the `Timer` linked list). This is best answered via profiling the code, so I've just chosen the simpler code for now. I also realized that if we want to avoid extraneous timer expiry events, there are simpler and more efficient ways than I am doing now (e.g., checking for timer updates after each input packet, not when each timer is set). All of this is explained in the comments above.

I believe that simplifying `mTimer` as you describe above also eliminates the edge cases you identified. I believe that the latest timer code is correct.

samkumar commented:
@abtink Thank you for clarifying the corner cases in the timer code. I updated the code to address those points. Let me know if there's anything else you'd like me to address.

samkumar commented:
I thought about this more and tested the updated timer code against the complete implementation (#6650), and I realized that there's one more edge case to handle. If a user callback is called, the user could mutate the linked list arbitrarily, by deinitializing and initializing endpoints other than the one that resulted in the callback. Thus, it's not enough to simply compute `next` at the start of the loop; we have to restart the traversal if a user callback is called. Note that this is rare, as most TCP callbacks just call TCP code and will not transfer control to the user. It's just one more edge case to handle. I've updated the code to that effect.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -166,15 +293,113 @@ Error Tcp::Listener::StopListening(void)
+    aTimer.GetInstance().Get<Tcp>().FireAllPendingTimers();
+}
+
+void Tcp::FireAllPendingTimers()
```
```suggestion
void Tcp::FireAllPendingTimers(void)
```

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -166,15 +293,113 @@ Error Tcp::Listener::StopListening(void)
+             * the linked list.
+             */
+            endpoint = mEndpoints.GetHead();
+            continue;
```
Should we also reset the `earliestPendingTimerExpiry` back to `now.GetDistantFuture()`?

We are going to go through the list again, and the value in `earliestPendingTimerExpiry ` that we are tracking at this point, may be from an end-point that is now removed from the list. So restting it makes sense to me.

samkumar replied:
Sure, we can do that as well. I restructured this using a `goto` to avoid duplicating initialization logic --- I know using `goto` is typically discouraged but in this case I really do think that it helps simplify things.

abtink commented on src/core/net/tcp6.cpp:
```diff
@@ -166,15 +293,113 @@ Error Tcp::Listener::StopListening(void)
+        {
+            pendingTimer               = true;
+            earliestPendingTimerExpiry = OT_MIN(earliestPendingTimerExpiry, expiry);
+        }
```
One idea/suggestion here which may make this part simpler:

Can we consider passing a reference to `earliestPendingTimerExpiry` directly to the `endpoint->FirePendingTimers()` for it to update it (if it wants)?
- I think this way we don't need `hasFutureTimer` and `expiry` variables and this block of code here (basically wecall  `FirePendingTimers(now, earliestPendingTimerExpiry)`).

Also a rename suggestion: How about if we rename it from `FirePendingTimers()` to `ProcessTimers()` (I am fine with existing name as well)

samkumar replied:
I agree that we can do this.

abtink reviewed (approved):
Thanks @samkumar . Looks great 👍 

Just couple of small suggestions below.

samkumar commented:
@abtink I addressed your latest round of comments. I also included the name change --- I changed `Tcp::FireAllPendingTimers` to `Tcp::ProcessTimers`. I left `Tcp::Endpoint::FirePendingTimers` as is because I didn't want to have two functions named `ProcessTimers` (though I admit the names were similar even before your suggestion).

jwhui commented:
> This pull request consists of three commits that add functionality in OpenThread needed for TCP and performs a small amount of initialization/deinitialization work in TCP itself. The purpose of this is to prepare for implementing TCP logic in future pull requests.

@samkumar , would you like to keep merge these commits separately or squash-merge? If so, can you append the PR # into the commit header? For example:

- `[tcp] make IPv6 header available to TCP segment processing routine (#6790)`

samkumar commented:
@jwhui Let's keep the commits separate. I've annotated each commit with the pull request number as you requested.
